File: sdk/python/aip/sequence.py
"""Per-stream sequence tracking (SPEC §5.4)."""


class Sequencer:
    """Assigns monotonically increasing seq per (session, source)."""

    def __init__(self) -> None:
        self._last: dict = {}

    def next(self, session: str, source: str) -> int:
        key = (session, source)
        self._last[key] = self._last.get(key, 0) + 1
        return self._last[key]

    def last(self, session: str, source: str) -> int:
        return self._last.get((session, source), 0)


class Receiver:
    """Classifies incoming messages against per-stream cursors (SPEC §5.4).

    Returns one of: "accept", "duplicate", "stale", "gap".
    """

    def __init__(self) -> None:
        self._applied_seq: dict = {}   # (session, source) -> last applied seq
        self._applied_count: dict = {}  # (session, source) -> messages applied
        self._ids: dict = {}            # (session, source) -> set of applied ids
        self.history: list = []         # applied (session, source, seq), in order

    def _key(self, msg) -> tuple:
        return (msg.session, msg.source)

    def classify(self, msg) -> str:
        key = self._key(msg)
        last = self._applied_seq.get(key, 0)
        if msg.seq == last + 1:
            return "accept"
        if msg.seq <= last:
            return "duplicate" if msg.id in self._ids.get(key, ()) else "stale"
        return "gap"

    def apply(self, msg) -> None:
        key = self._key(msg)
        self._applied_seq[key] = msg.seq
        self._applied_count[key] = self._applied_count.get(key, 0) + 1
        self._ids.setdefault(key, set()).add(msg.id)
        self.history.append((msg.session, msg.source, msg.seq))

    def cursor(self, session: str, source: str) -> int:
        return self._applied_seq.get((session, source), 0)

    def cursors(self, session: str) -> dict:
        return {s: seq for (sess, s), seq in self._applied_seq.items() if sess == session}

    def applied_count(self, session: str, source: str) -> int:
        return self._applied_count.get((session, source), 0)

File: sdk/typescript/src/sequence.ts
/**
 * Per-stream sequence tracking (SPEC §5.4).
 */
import type { Message } from "./messages.js";

export type Classification = "accept" | "duplicate" | "stale" | "gap";

export class Sequencer {
  private last = new Map<string, number>();

  private key(session: string, source: string): string {
    return `${session}\u0000${source}`;
  }

  next(session: string, source: string): number {
    const k = this.key(session, source);
    const n = (this.last.get(k) ?? 0) + 1;
    this.last.set(k, n);
    return n;
  }

  lastOf(session: string, source: string): number {
    return this.last.get(this.key(session, source)) ?? 0;
  }
}

export class Receiver {
  private appliedSeq = new Map<string, number>();
  private appliedCount = new Map<string, number>();
  private ids = new Map<string, Set<string>>();
  history: Array<[string, string, number]> = [];

  private key(m: Message): string {
    return `${m.session}\u0000${m.source}`;
  }

  classify(m: Message): Classification {
    const k = this.key(m);
    const last = this.appliedSeq.get(k) ?? 0;
    if (m.seq === last + 1) return "accept";
    if (m.seq <= last) {
      return this.ids.get(k)?.has(m.id) ? "duplicate" : "stale";
    }
    return "gap";
  }

  apply(m: Message): void {
    const k = this.key(m);
    this.appliedSeq.set(k, m.seq);
    this.appliedCount.set(k, (this.appliedCount.get(k) ?? 0) + 1);
    if (!this.ids.has(k)) this.ids.set(k, new Set());
    this.ids.get(k)!.add(m.id);
    this.history.push([m.session, m.source, m.seq]);
  }

  cursor(session: string, source: string): number {
    return this.appliedSeq.get(`${session}\u0000${source}`) ?? 0;
  }

  cursors(session: string): Record<string, number> {
    const out: Record<string, number> = {};
    for (const [k, seq] of this.appliedSeq) {
      const [sess, source] = k.split("\u0000");
      if (sess === session) out[source] = seq;
    }
    return out;
  }

  appliedCountOf(session: string, source: string): number {
    return this.appliedCount.get(`${session}\u0000${source}`) ?? 0;
  }
}